sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificMutableRow.scala

```diff
@@ -230,13 +251,17 @@ final class SpecificMutableRow(val values: Array[MutableValue]) extends MutableR
     new GenericRow(newValues)
   }
 
-  override def update(ordinal: Int, value: Any): Unit = {
-    if (value == null) setNullAt(ordinal) else values(ordinal).update(value)
+  override def update(ordinal: Int, value: Any): Unit = value match {
+    case null => setNullAt(ordinal)
+    case s: String =>
+      // for tests
+      throw new Exception("String should be converted into UTF8String")
+    case other => values(ordinal).update(value)
```

Comment:
I'm not actually sure, but I wonder if it would be faster to do a plain null check followed by an asInstanceOf, rather than pattern matching.  You will still get a class cast exception in tests if there is a mistake.